Add resource layout scheme

AMD-dwang opened the pull request:
Indirect scheme is designed for the case that the pipeline layout only
contains part of information of the final executable pipeline. So that
user data is used in conservative way to make sure that this pipeline
layout is alaways compatiable with other layouts which may contain the
descriptor which is not known currently.

In indirect mode, the user data registers for various resources are
allocated in the following fashion:
 1. one user data entry for vertex buffer
 2. one user data entry for the push constant buffer
 3. descriptor set index for each set.

amdvlk-admin commented:
Can one of the admins verify this patch?

AMD-dwang commented:
Can some one take a look?

kuhar commented on include/vkgcDefs.h:
```diff
@@ -378,6 +379,14 @@ enum class ThreadGroupSwizzleMode : unsigned {
+enum class ResourceMappingLayout : unsigned {
+  // Compact scheme make full use of all the user data registers.
+  Compact = 0,
+  // The searching path of resource belongs to a specific binding is fixed in indirect scheme.
```
nit: The search path?

AMD-dwang replied:
This comment is copied from driver, but it is really hard to understand.

kuhar commented on include/vkgcDefs.h:
```diff
@@ -378,6 +379,14 @@ enum class ThreadGroupSwizzleMode : unsigned {
 
+/// Represents mapping layout of the resources used in shaders
+enum class ResourceMappingLayout : unsigned {
+  // Compact scheme make full use of all the user data registers.
```
nit: can you put these comments on the right of values, just like in the lgc enum?

AMD-dwang replied:
Yes, I will change it . Thanks.

kuhar commented on lgc/patch/PatchEntryPointMutate.cpp:
```diff
@@ -1190,6 +1190,62 @@ void PatchEntryPointMutate::addUserDataArgs(SmallVectorImpl<UserDataArg> &userDa
     return;
   }
 
+  // Define the function
```
This comment is not very helpful. It'd be better to comment what this lambda is used for, why do we need it, etc.

kuhar commented on lgc/patch/PatchEntryPointMutate.cpp:
```diff
@@ -1190,6 +1190,62 @@ void PatchEntryPointMutate::addUserDataArgs(SmallVectorImpl<UserDataArg> &userDa
   }
 
+  // Define the function
+  auto addPushConstArgs = [&](const ResourceNode &node) {
```
This lambda is very long. How about making a freestanding or a member function?

AMD-dwang replied:
I make it a member function. Thanks.

kuhar commented on lgc/patch/PatchEntryPointMutate.cpp:
```diff
@@ -1190,6 +1190,62 @@ void PatchEntryPointMutate::addUserDataArgs(SmallVectorImpl<UserDataArg> &userDa
+      }
+    } else {
+      // Mark push constant for spill for compute library.
+      userDataUsage->pushConstSpill = true;
```
nit: Can you flip the if condition and put this `else` case first? The current `then` case is very long and it's not immediately obvious which `if` this `else` belongs to.

kuhar commented on llpc/test/shaderdb/PipelineVsFs_TestNonRootPushConst.pipe:
```diff
@@ -0,0 +1,171 @@
+
+
```
Can you describe what the test checks?

kuhar commented on llpc/test/shaderdb/PipelineVsFs_TestNonRootPushConst.pipe:
```diff
@@ -0,0 +1,171 @@
+
+; BEGIN_SHADERTEST
+; RUN: amdllpc -spvgen-dir=%spvgendir% -v %gfxip %s | FileCheck -check-prefix=SHADERTEST %s
+; SHADERTEST-LABEL: {{^// LLPC}} SPIRV-to-LLVM translation results
+; SHADERTEST: AMDLLPC SUCCESS
```
This does not match any compilation results (LLVM IR, GCN, Pal metadata, etc.). How can we tell if the code does what its supposed to?

AMD-dwang replied:
It will crash when compiling this pipeline without this change.  This pipeline is aim to verify this change, not to special instructions. It is okay if it can be compiled successfully. Thanks.

kuhar replied:
Can you add a comment with this explanation?

AMD-dwang replied:
I added it. Thanks.

kuhar replied:
I don't see the checks that Steven suggested in the test, so I'm marking this as not resolved yet.

AMD-dwang replied:
Currently, are there such three test cases with compact mode?

s-perron replied:
I don't know if they currently exist or not.  I believe the test that you have here would normally not spill the push constant, but you can look into that.  If it does spill you can probably change the type of mvp to float, and it will unspill it.  I don't know if there is a case for the compute shaders, but it should be easy enough to write.  I wrote a test for relocatable shaders that checks that it works when the push const is spilled: 

https://github.com/GPUOpen-Drivers/llpc/blob/b780577b54bcccb85c191d022b63ea5ce9cc72a4/llpc/test/shaderdb/relocatable_shaders/PipelineCs_PushConst.pipe.

kuhar commented on llpc/test/shaderdb/PipelineVsFs_TestNonRootPushConst.pipe:
```diff
@@ -0,0 +1,171 @@
+attribute[1].location = 1
+attribute[1].binding = 0
+attribute[1].format = VK_FORMAT_R32G32B32A32_SFLOAT
+attribute[1].offset = 16
```
missing newline

kuhar reviewed (commented):
I only commented on some cosmetics. Steven would be a better person to check the logic.

s-perron commented on lgc/patch/PatchEntryPointMutate.cpp:
```diff
@@ -1325,6 +1281,69 @@ void PatchEntryPointMutate::addUserDataArg(SmallVectorImpl<UserDataArg> &userDat
+// @param node : PushConst resource node
+// @param builder : IRBuilder (just for getting types)
+void PatchEntryPointMutate::addPushConstArgs(SmallVectorImpl<UserDataArg> &userDataArgs, const ResourceNode &node,
+                                             llvm::IRBuilder<> &builder) {
```
It looks a little suspicious to me that you can just use the same code here regardless of whether the node is in the root table or a descriptor table.  I expect that you will need to different function that will share some similar code.

AMD-dwang replied:
> To make sure that the implementation actually works, you could also add an amber testcase like the one for a16 (https://github.com/GPUOpen-Drivers/llpc/blob/dev/test/amber/a16.amber). It may be a bit hidden in the a16 example, but they support lit CHECK lines to check for lines in the pipeline dumps.
> 
> (These tests are not yet run in CI unfortunately, but you can run them locally with `ninja check-amber` in an xgl build.)

Thanks, I try to add it.

s-perron commented on lgc/patch/PatchEntryPointMutate.cpp:
```diff
@@ -1325,6 +1281,69 @@ void PatchEntryPointMutate::addUserDataArg(SmallVectorImpl<UserDataArg> &userDat
+
+  // Ensure we mark the push constant's part of the spill table as used.
+  if (userDataUsage->pushConstSpill)
+    userDataUsage->spillUsage = std::min(userDataUsage->spillUsage, node.offsetInDwords);
```
Is this right?  Should we be spilling the root table even if `node` is not in the root table?

AMD-dwang replied:
I changed "node.offsetInDwords" to "tableNode.offsetInDwords". Thanks.

s-perron commented on lgc/patch/PatchEntryPointMutate.cpp:
```diff
@@ -1325,6 +1281,69 @@ void PatchEntryPointMutate::addUserDataArg(SmallVectorImpl<UserDataArg> &userDat
+
+      // Add the arg (part of the push const) that we can potentially unspill.
+      addUserDataArg(userDataArgs, node.offsetInDwords + dwordOffset, pushConstOffset.dwordSize,
+                     "pushConst" + Twine(dwordOffset), &pushConstOffset.entryArgIdx, builder);
```
I would expect that in the indirect case, these values will always have to be unspilled.  Can you point out the code that will do the unspilling?

s-perron replied:
Sorry, I think I meant that it should never be unspilled.  Are we sure that will not happen?

AMD-dwang replied:
The code is the same as before. I only move it. I think it should be correct, after all it has been stable. Thanks.

s-perron replied:
Even if this is the same code that does not mean it is correct because you are executing it in a different context.  For example, should you be setting `userDataUsage->spillUsage` below if the push constant node is not in the root table?

I still wonder what will happen with the pushConst argument in `PatchEntryPointMutate::determineUnspilledUserDataArgs`.  Again, that code was written assuming that push constants are in the root table.  What will that code do if it decides pass the push constant in SGPRs instead of loading it from memory?  I assume that should never happen, but no code was added to make sure it does not.

s-perron reviewed (changes requested):
I have concerns about the correctness in all cases.  If I saw tests that called out what instructions are expected and what the PAL metadata is suppose to be, I could be convinced that this is correct.

Flakebi commented:
To make sure that the implementation actually works, you could also add an amber testcase like the one for a16 (https://github.com/GPUOpen-Drivers/llpc/blob/dev/test/amber/a16.amber).
It may be a bit hidden in the a16 example, but they support lit CHECK lines to check for lines in the pipeline dumps.

(These tests are not yet run in CI unfortunately, but you can run them locally with `ninja check-amber` in an xgl build.)

AMD-dwang commented on lgc/interface/lgc/CommonDefs.h:
```diff
@@ -89,6 +89,12 @@ enum class ResourceNodeType : unsigned {
+// Represents mapping layout of the resources used in shaders
+enum class ResourceMappingLayout : unsigned {
+  Compact = 0, /// Compact scheme make full use of all the user data registers.
+  Indirect     /// The searching path of resource belongs to a specific binding is fixed in indirect scheme.
```
The same enumeration was added in driver, which changes the resource mapping layout. I think it's best to keep their names and comments consistent. Also this enumeration is not only for push Constant, the whole resource mapping is affected. Thanks.

s-perron replied:
At the very least, could up update the comment above, even if you can't update the name.  If I am reading code that checks if the layout is indirect, I have no way of knowing what that means because the comment references itself: what is search path for the?  The comment for indirect does not read well.  I'm guessing "belongs" is a typo and should be "belonging".

Can we ask the driver team to update the comment in the driver code?  Then you can keep them the same.

AMD-dwang replied:
"The fixed user data registers. When the pipeline layout only contains part of layout information of the final executable pipeline, we don't know the offset of user data register, so we must use fixed offset." 

Is it feasible to change the comment like above?

AMD-dwang commented on llpc/test/shaderdb/PipelineVsFs_TestNonRootPushConst.pipe:
```diff
@@ -0,0 +1,171 @@
+options.adjustDepthImportVrs = 0
+options.disableLicmThreshold = 0
+options.unrollHintThreshold = 0
+options.dontUnrollHintThreshold = 0
```
Thanks, I will remove unnecessary options.

AMD-dwang commented:
The change of driver has been submitted, an amber test is not be added.

AMD-dwang commented:
@s-perron I have addressed the comments. this change only adds a resource mapping mode,  no other feature is changed. Please help to take a look again. Thank you.

kuhar commented:
> @s-perron I have addressed the comments. this change only adds a resource mapping mode, no other feature is changed. Please help to take a look again. Thank you.

@AMD-dwang Steven is on vacation and will be back in January.

kuhar reviewed (changes requested):
There's one outstanding change request from Steven's previous review: https://github.com/GPUOpen-Drivers/llpc/pull/1587/files#r769648173

JaxLinAMD commented:
/rebase

amdvlk-admin commented:
### Test summary for commit 5c010407a433f6aea2f06f5ec1a3086ddcf160bd
<details><summary>CTS tests (Failed: 0/187509)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 36142/64680 (55.9%)</li>
<li>Failed: 0/64680 (0.0%)</li>
<li>Not Supported: 28538/64680 (44.1%)</li><li>Warnings: 0/64680 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31577/58149 (54.3%)</li>
<li>Failed: 0/58149 (0.0%)</li>
<li>Not Supported: 26572/58149 (45.7%)</li><li>Warnings: 0/58149 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37446/64680 (57.9%)</li>
<li>Failed: 0/64680 (0.0%)</li>
<li>Not Supported: 27234/64680 (42.1%)</li><li>Warnings: 0/64680 (0.0%)</li>
</ul></details>


</ul></details>

s-perron reviewed (dismissed):
The tests needs to be expanded.  I'd also like some better documentation on the "compact" and "indirect" schemes.  We should also have an separate discussion on why we need this.  I don't see the reason for this change in the discussion.  I'll send an email about it.

s-perron commented on tool/dumper/vkgcPipelineDumper.cpp:
```diff
@@ -684,6 +685,7 @@ void PipelineDumper::dumpComputeStateInfo(const ComputePipelineBuildInfo *pipeli
 void PipelineDumper::dumpPipelineOptions(const PipelineOptions *options, std::ostream &dumpFile) {
   dumpFile << "options.includeDisassembly = " << options->includeDisassembly << "\n";
   dumpFile << "options.scalarBlockLayout = " << options->scalarBlockLayout << "\n";
+  dumpFile << "options.resourceLayoutScheme = " << options->resourceLayoutScheme << "\n";
```
I think this file should also include the resource layout scheme in the hash.

AMD-dwang replied:
Oh, I think hash includes resource mapping, so I didn't add it.  After checking again, it only is added when relocatable is false. Thanks, I will add it.

AMD-dwang commented on lgc/builder/DescBuilder.cpp:
```diff
@@ -221,13 +221,27 @@ Value *DescBuilder::CreateGetDescPtr(ResourceNodeType descType, unsigned descSet
+    Value *highHalf = getInt32(HighAddrPc);
+    Value *descPtr = CreateNamedCall(
+        lgcName::DescriptorTableAddr, getInt8Ty()->getPointerTo(ADDR_SPACE_CONST),
+        {getInt32(unsigned(ResourceNodeType::PushConst)), getInt32(subNode.set), getInt32(subNode.binding), highHalf},
```
Here set and binding are a fixed value, set = -1,binding  = 0, they are not meaningful values,  but we can find matched resource node according to them.

AMD-dwang commented on llpc/test/shaderdb/general/PipelineVsFs_TestIndirectResourceLayout.pipe:
```diff
@@ -0,0 +1,69 @@
+; SHADERTEST: call [16 x i8] addrspace(4)* (...) @lgc.create.load.push.constants.ptr.p4a16i8()
+; SHADERTEST-LABEL: define dllexport spir_func void @lgc.shader.FS.main()
+; SHADERTEST: call i8 addrspace(4)* @lgc.descriptor.table.addr(i32 9, i32 -1, i32 0, i32 -1)
+; SHADERTEST: AMDLLPC SUCCESS
```
This test is not to verify the use of push constant variable, I just want to show that we will call `@lgc.descriptor.table.addr` to load descriptor in indirect scheme.  The second call is part of the 'before-patching results', I will fix here.
`@lgc.descriptor.table.addr(i32 9, i32 -1, i32 0, i32 -1)` can represent we are using `@lgc.descriptor.table.addr` to load push constant descriptor, other instructions are not important.
Thanks.

s-perron replied:
I still think the test should be checking for more, but ,if other reviewers do not think it is necessary, I won't stand in the way.

AMD-dwang replied:
Okay, I will add more instructions. Thanks

AMD-dwang commented on llpc/context/llpcPipelineContext.cpp:
```diff
@@ -334,6 +334,7 @@ void PipelineContext::setOptionsInPipeline(Pipeline *pipeline, Util::MetroHash64
   options.disableImageResourceCheck = getPipelineOptions()->disableImageResourceCheck;
   options.enableInterpModePatch = getPipelineOptions()->enableInterpModePatch;
   options.pageMigrationEnabled = getPipelineOptions()->pageMigrationEnabled;
+  options.resourceLayoutScheme = static_cast<lgc::ResourceLayoutScheme>(getPipelineOptions()->resourceLayoutScheme);
```
Yes, I will add an option. Thanks.

AMD-dwang commented on include/vkgcDefs.h:
```diff
@@ -393,6 +394,12 @@ enum class ThreadGroupSwizzleMode : unsigned {
+/// Represents mapping layout of the resources used in shaders
+enum class ResourceLayoutScheme : unsigned {
+  Compact = 0, ///< Compact scheme make full use of all the user data registers.
+  Indirect     ///< Fixed layout, push constant will be the sub node of DescriptorTableVaPtr
```
When VK_PIPELINE_LAYOUT_CREATE_INDEPENDENT_SETS_BIT_EXT is used, we will using indirect mode.

s-perron replied:
Maybe a full document is not needed, but at least move the comment you have the EntryPointMutate.cpp here.  This will make it easier to find.  If someone wants to find out what the indirect scheme is, they look at the definition of the enum.

AMD-dwang commented:
Add an option "resource-layout-scheme" to support 'offline' compilation

amdvlk-admin commented:
### Test summary for commit 20deb2dea6c352cfeeef04903d2e5d758a26b98a
<details><summary>CTS tests (Failed: 0/187823)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 36645/65226 (56.2%)</li>
<li>Failed: 0/65226 (0.0%)</li>
<li>Not Supported: 28581/65226 (43.8%)</li><li>Warnings: 0/65226 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31111/57371 (54.2%)</li>
<li>Failed: 0/57371 (0.0%)</li>
<li>Not Supported: 26260/57371 (45.8%)</li><li>Warnings: 0/57371 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37804/65226 (58.0%)</li>
<li>Failed: 0/65226 (0.0%)</li>
<li>Not Supported: 27422/65226 (42.0%)</li><li>Warnings: 0/65226 (0.0%)</li>
</ul></details>


</ul></details>

amdvlk-admin commented:
### Test summary for commit 66ad702cc9a0bb97a55ceafcc8e84d4ff8f6d567
<details><summary>CTS tests (Failed: 0/187823)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 36645/65226 (56.2%)</li>
<li>Failed: 0/65226 (0.0%)</li>
<li>Not Supported: 28581/65226 (43.8%)</li><li>Warnings: 0/65226 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31111/57371 (54.2%)</li>
<li>Failed: 0/57371 (0.0%)</li>
<li>Not Supported: 26260/57371 (45.8%)</li><li>Warnings: 0/57371 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37804/65226 (58.0%)</li>
<li>Failed: 0/65226 (0.0%)</li>
<li>Not Supported: 27422/65226 (42.0%)</li><li>Warnings: 0/65226 (0.0%)</li>
</ul></details>


</ul></details>

s-perron commented on lgc/patch/PatchEntryPointMutate.cpp:
```diff
@@ -364,6 +364,24 @@ void PatchEntryPointMutate::gatherUserDataUsage(Module *module) {
+            // 1. one user data entry for vertex buffer
+            // 2. one user data entry for the push constant buffer
+            // 3. descriptor set index for each set
+            // If the type is pushconst, the index is set 0, others are set + 1.
```
This type of comment should be somewhere more discoverable.  If I am reading `DescBuilder.cpp` I might want this information.  Could you place this with the definition of `ResourceLayoutScheme` in include/vkgcDefs.h.

kuhar commented on llpc/test/shaderdb/general/PipelineVsFs_TestIndirectResourceLayout.pipe:
```diff
@@ -0,0 +1,69 @@
+; Check the resource layout mode. In "Indirect" mode, "PushConst" resource is not in root node.
+
+; BEGIN_SHADERTEST
+; RUN: amdllpc -spvgen-dir=%spvgendir% -v %gfxip %s | FileCheck -check-prefix=SHADERTEST %s
```
I'm a bit confused. I don't see `--resource-layout-scheme` specified here, doesn't  this test default to the compact mode then?

kuhar commented on lgc/builder/DescBuilder.cpp:
```diff
@@ -221,13 +221,35 @@ Value *DescBuilder::CreateGetDescPtr(ResourceNodeType descType, unsigned descSet
-  std::string callName = lgcName::PushConst;
-  addTypeMangling(returnTy, {}, callName);
-  return CreateNamedCall(callName, returnTy, {}, Attribute::ReadOnly, instName);
+  bool isIndirect = getPipelineState()->getOptions().resourceLayoutScheme == ResourceLayoutScheme::Indirect;
```
nit: add `const` if this won't be modified later in the function

kuhar commented on lgc/builder/DescBuilder.cpp:
```diff
@@ -221,13 +221,35 @@ Value *DescBuilder::CreateGetDescPtr(ResourceNodeType descType, unsigned descSet
+        {getInt32(unsigned(ResourceNodeType::PushConst)), getInt32(subNode.set), getInt32(subNode.binding), highHalf},
+        Attribute::ReadNone);
+    return CreateBitCast(descPtr, returnTy);
+  } else {
```
nit: else after return: https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

amdvlk-admin commented:
### Test summary for commit f38083652b54d78e7b6e3513fea7917db0fc05da
<details><summary>CTS tests (Failed: 0/186422)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 36645/65226 (56.2%)</li>
<li>Failed: 0/65226 (0.0%)</li>
<li>Not Supported: 28581/65226 (43.8%)</li><li>Warnings: 0/65226 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 30002/55970 (53.6%)</li>
<li>Failed: 0/55970 (0.0%)</li>
<li>Not Supported: 25968/55970 (46.4%)</li><li>Warnings: 0/55970 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37804/65226 (58.0%)</li>
<li>Failed: 0/65226 (0.0%)</li>
<li>Not Supported: 27422/65226 (42.0%)</li><li>Warnings: 0/65226 (0.0%)</li>
</ul></details>


</ul></details>

AMD-dwang commented:
Do someone have any comments for this change? It is blocking some tests. Thanks.

amdrexu commented:
We have some internal test failures that depend on this fix. If the PR could be reviewed and gets merged timely, that will be fine for us.

s-perron reviewed (commented):
LGTM
